[Zeppelin-4049] Hadoop Submarine (Machine Learning) interpreter

liuxunorg opened the pull request:
### What is this PR for?
Added the new submarine project for Hadoop in zeppelin to implement machine learning algorithm development and task scheduling in zeppelin.
1. Supports algorithm development and debugging of tensorflow
2. Support offline training of tensorflow
3. Support for Docker containers with YARN-3.1

### What type of PR is it?
[Feature]

### Todos
* [ ] - Task

### What is the Jira issue?
* https://issues.apache.org/jira/browse/ZEPPELIN-4049

### How should this be tested?

* Doc: https://github.com/hadoopsubmarine/hadoop-submarine-ecosystem/tree/master/submarine-integration-zeppelin

* Youtebe：https://www.youtube.com/channel/UC4JBt8Y8VJ0BW0IM9YpdCyQ

### Screenshots (if appropriate)
![Submarine-Dashboard](https://user-images.githubusercontent.com/3677382/54174286-a722dc00-44c0-11e9-850e-08c247761472.gif)

[CI Pass](https://travis-ci.org/liuxunorg/zeppelin/builds/505020752)

### Questions:
* Does the licenses files need update? No
* Is there breaking changes for older versions? No
* Does this needs documentation? Yes

cquptEthan commented:
LGTM

zjffdu commented:
Thanks @liuxunorg  It is a very interested interpreter for zeppelin in deep learning area. Since it is a very PR, could you send one mail in user mail list to gather more attention on this feature ? I hope more people could try it and gave feedback.

zjffdu commented on submarine/pom.xml:
```diff
@@ -0,0 +1,159 @@
+  <properties>
+    <!--library versions-->
+    <interpreter.name>submarine</interpreter.name>
+    <hadoop.version>2.7.3</hadoop.version>
```
Do I need to change it 3.x if I want to use submarine in hadoop 3 ?

liuxunorg replied:
Here Hadoop is used to upload algorithm files to HDFS, Hadoop3.1+ needs to be deployed locally like a spark. I have instructions in the `interpreter/submarine.md` documentation.
Submarine is developing a client that does not rely on hadoop, There is only one JAR file package. I will update it when it is released.

zjffdu commented on submarine/src/main/java/org/apache/zeppelin/submarine/PySubmarineInterpreter.java:
```diff
@@ -0,0 +1,84 @@
+  public InterpreterResult interpret(String st, InterpreterContext context)
+      throws InterpreterException {
+    // algorithm & checkpoint path support replaces ${username} with real user name
+    String algorithmPath = properties.getProperty(
```
Should this piece of code be applied in IPySubmarineInterpreter ? IIUC, IPySubmarineInterpreter should behave the same as PySubmarineInterpreter except more ipython features

liuxunorg replied:
Now our users are using `python` to develop machine learning algorithms. Because submarine now offers online services, And have a certain amount of code, So I hope I can merge this phase code first. Subsequent improvements to ipython support.

zjffdu commented:
@liuxunorg Is it possible to add integration test for this interpreter ? E.g. Run one deep learning task via SubmarineInterpreter ? I understand it may be not possible because currently submarine is not mature enough, just want to bring this out if it is possible.

liuxunorg commented:
I also thought about the way to provide integration testing.
However, machine learning requires strict environmental requirements.
For example, need to install `docker`, `tensorflow`, and `YARN` in test environment.
This requires a process of gradual improvement.
Therefore, only documents and operation videos can be provided at present.

Doc: https://github.com/hadoopsubmarine/hadoop-submarine-ecosystem/tree/master/submarine-integration-zeppelin
Youtebe：https://www.youtube.com/channel/UC4JBt8Y8VJ0BW0IM9YpdCyQ

zjffdu commented:
@liuxunorg Regarding integration, tensorflow can be installed in travis. (See https://github.com/apache/zeppelin/blob/master/testing/install_external_dependencies.sh)
And We can also set up a mini yarn cluster (See https://github.com/apache/zeppelin/blob/master/zeppelin-interpreter-integration/src/test/java/org/apache/zeppelin/integration/MiniHadoopCluster.java)
For docker, I am not sure whether travis can install it, need to investigate more.

felixcheung commented on scripts/docker/submarine/submarine-interpreter-cpu-1.0.0/Dockerfile:
```diff
@@ -0,0 +1,19 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM zeppelin-tensorflow_1.10-hadoop_3.1.2-cpu:1.0.0
```
how is zeppelin-tensorflow_1.10-hadoop_3.1.2-cpu created? it looks like it is built by a script in the PR, how is this name chosen?

liuxunorg replied:
1. I have carefully considered the issue of the docker name. Otherwise it will be difficult to maintain in the future.
2. Because the name needs to include the name and version number of the main component in the image.
3. **Naming rules**

    + The name `_` indicates the separator, splitting different systems, for example: `tensorflow` and `hadoop`
    + The name`-` indicates the connector, the connection system and the version number, for example: `hadoop-3.1.2`

felixcheung replied:
if we are not releasing / publishing this docker image as a part of the Apache Zeppelin release, I think this is ok, otherwise a different naming will need to be used

felixcheung replied:
why is cpu at the end? maybe it should be `zeppelin-cpu-tensorflow_1.10-hadoop_3.1.2`?

also any reason for tensorflow 1.10 vs the latest 1.13? https://github.com/tensorflow/tensorflow/releases/tag/v1.13.1

liuxunorg replied:
About the Docker Image Name, Can you give me some suggestion? Thanks!

liuxunorg replied:
Yes, tensorflow uses version 1.13.0. I made a mistake and I will correct this error.

felixcheung replied:
hmm, it looks to be uncommon to put version number in image name? https://hub.docker.com/u/apache?page=2

other than that I don't have any good/useful suggestion. perhaps the latest you have is fine

felixcheung commented on scripts/docker/submarine/tensorflow_1.10-hadoop_3.1.2-gpu-1.0.0/Dockerfile:
```diff
@@ -0,0 +1,71 @@
+
+# Install hadoop
+ENV HADOOP_VERSION="3.1.2"
+RUN wget http://mirrors.hust.edu.cn/apache/hadoop/common/hadoop-${HADOOP_VERSION}/hadoop-${HADOOP_VERSION}.tar.gz
```
we can't hardcode the mirror

liuxunorg replied:
1. There are some bugs in the tensorflow version for distributed training.
2. Tensorflow also has a lot of problems with different versions of hadoop compatibility.

Now, the docker we provide is a companion version that has been tested in the production environment.
Therefore, the version number cannot be modified at will.

felixcheung replied:
I don't mean the hadoop version - I mean the "http://mirrors..."  site. 
a) this is known to be unreliable
b) as a policy, all apache download should be from the download script

https://www.apache.org/dyn/closer.cgi/hadoop/common/hadoop-3.1.2/hadoop-3.1.2-src.tar.gz

https://hadoop.apache.org/releases.html

liuxunorg replied:
OK, Thanks for your correction. :-)

felixcheung commented on scripts/docker/submarine/tensorflow_1.10-hadoop_3.12-cpu-1.0.0/Dockerfile:
```diff
@@ -0,0 +1,69 @@
+
+# Install hadoop
+ENV HADOOP_VERSION="3.1.2"
+RUN wget http://mirrors.hust.edu.cn/apache/hadoop/common/hadoop-${HADOOP_VERSION}/hadoop-${HADOOP_VERSION}.tar.gz
```
ditto

liuxunorg replied:
The version number cannot be modified.
ditto.

felixcheung commented on submarine/src/main/java/org/apache/zeppelin/submarine/SubmarineInterpreter.java:
```diff
@@ -0,0 +1,275 @@
+    }
+    if (needUpdateConfig) {
+      needUpdateConfig = false;
+      if (currentReplName.equals("submarine") || currentReplName.isEmpty()) {
```
IIRC, the interpreter name is configurable? eg.  spark is `spark2` in some config. so likely you shouldn't hard code `submarine`

liuxunorg replied:
I am at https://github.com/apache/zeppelin/pull/3271 , Has all interpreters been able to customize some configurations.
Add this code to the submarine interpreter for a better experience.
Will not have any effect on other interpreters.

felixcheung replied:
I mean like install any interpreter with a different name `--name` in
https://zeppelin.apache.org/docs/0.8.0/usage/interpreter/installation.html#spark-interpreter-for-scala-210-and-211

or just name it something different - the "Name" box can say "sub" and the "Interpreter" box can be "submarine"
https://zeppelin.apache.org/docs/0.8.0/usage/interpreter/overview.html#what-is-zeppelin-interpreter

liuxunorg replied:
Ok, I am going to change the way to determine if it is a submarine interpreter.

felixcheung commented on submarine/src/main/java/org/apache/zeppelin/submarine/SubmarineInterpreter.java:
```diff
@@ -0,0 +1,275 @@
+
+  @Override
+  public int getProgress(InterpreterContext context) {
+    return 0;
```
can this support progress in the future?

liuxunorg replied:
The submarine interpreter can display the execution status of the task more abundantly.
In addition, I discussed with Jeff that this richer UI can be packaged into components for use by other interpreters.

![JobRun](https://user-images.githubusercontent.com/3677382/54327310-9d29e600-4644-11e9-87e6-51d2421a3e75.gif)

felixcheung replied:
thats cool

felixcheung commented on submarine/src/main/java/org/apache/zeppelin/submarine/SubmarineShellInterpreter.java:
```diff
@@ -0,0 +1,107 @@
+  protected boolean isKerboseEnabled() {
+    if (!StringUtils.isAnyEmpty(getProperty(ZEPPELIN_SUBMARINE_AUTH_TYPE))
+        && getProperty(ZEPPELIN_SUBMARINE_AUTH_TYPE)
+        .equalsIgnoreCase("kerberos")) {
```
indentation? can you merge this into the last line

liuxunorg replied:
OK.

felixcheung commented on submarine/src/main/java/org/apache/zeppelin/submarine/commons/SubmarineConstants.java:
```diff
@@ -0,0 +1,119 @@
+      = "machinelearing.distributed.enable";
+
+  public static final String ZEPPELIN_INTERPRETER_RPC_PORTRANGE
+      = "zeppelin.interpreter.rpc.portRange";
```
is this defined else where?

liuxunorg replied:
This definition is used in the second PR (`YarnInterpreterLauncher`), This PR is merged and will be submitted shortly.
1. The `submarine interpreter` runs in a remote Docker container via `YARN`.
2. Work together through this PRC port with the local zeppelin server.

felixcheung replied:
if this only affects submarine, suggest calling it `submarine.rpc.portRange` or `submarine.interpreter.rpc.portRange`

liuxunorg replied:
Done.

felixcheung commented on submarine/src/main/java/org/apache/zeppelin/submarine/commons/SubmarineConstants.java:
```diff
@@ -0,0 +1,119 @@
+  public static final String PS_LAUNCH_CMD = "PS_LAUNCH_CMD";
+  public static final String WORKER_LAUNCH_CMD = "WORKER_LAUNCH_CMD";
+  public static final String MACHINELEARING_DISTRIBUTED_ENABLE
+      = "machinelearing.distributed.enable";
```
`machinelearing` -> `machinelearning`

liuxunorg replied:
Done.

liuxunorg commented:
@zjffdu , I am improving the code based on comments.
1. Added Tensorflow installation in travis CI, 
2. Added Tensorflow in submarine test case, Support any version of tensorflow.
3. Added SubmarineInterpreterTest.
4. Added PySubmarineInterpreterTest.

CI pass https://travis-ci.org/liuxunorg/zeppelin/builds/507793860. 
Please help me review the code. Thanks!

zjffdu commented on .travis.yml:
```diff
@@ -89,7 +89,7 @@ matrix:
     - jdk: "oraclejdk8"
       dist: trusty
-      env: PYTHON="3" SPARKR="true" SCALA_VER="2.10" PROFILE="-Pscala-2.10" BUILD_FLAG="install -DskipTests -DskipRat" TEST_FLAG="test -DskipRat" MODULES="-pl $(echo .,zeppelin-interpreter,zeppelin-interpreter-api,${INTERPRETERS} | sed 's/!//g')" TEST_PROJECTS=""
+      env: PYTHON="3" SPARKR="true" SCALA_VER="2.10" TENSORFLOW="1.0.0" PROFILE="-Pscala-2.10" BUILD_FLAG="install -DskipTests -DskipRat -am" TEST_FLAG="test -DskipRat" MODULES="-pl $(echo .,zeppelin-interpreter,zeppelin-interpreter-api,${INTERPRETERS} | sed 's/!//g')" TEST_PROJECTS=""
```
Is tensorflow 1.0.0 the latest version that submarine supports ?

liuxunorg replied:
In the PySubmarineInterpreterTest case `PySubmarineInterpreterTest::testTensorflow()`, just import the tensorflow library, then call tensorflow to get the version number function.
If PySubmarineInterpreterTest fails, an exception is thrown. So the test case has nothing to do with the version of tensorflow.
Official use is through `scripts/docker/submarine/1.0.0/zeppelin-cpu-tensorflow_1.13.1-hadoop_3.1.2/Dockerfile`.

zjffdu commented on scripts/docker/submarine/1.0.0/zeppelin-cpu-submarine-interpreter/Dockerfile:
```diff
@@ -0,0 +1,20 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+FROM zeppelin-cpu-tensorflow_1.13.0-hadoop_3.1.2:1.0.0
```
Where is it ? I could not find it in dockerhub

zjffdu replied:
Sorry, I see it below of this PR

liuxunorg replied:
`Zeppelin-cpu-tensorflow_1.13.0-hadoop_3.1.2` Is the basic image that we need to provide,
Push to dockerhub for users to use, So now there is no dockerhub inside.
The docker images for `zeppelin-cpu-tensorflow_1.13.0-hadoop_3.1.2` are in `scripts/docker/submarine/1.0.0/zeppelin-cpu-tensorflow_1.13.1-hadoop_3.1.2/Dockerfile`.
Regarding the name of the image, felixcheung also made some comments.
I want to do this:
#### Base Docker Image:(Used to run the tensorflow algorithm in yarn)
```
zeppelin hadoop submarine cpu base image name: zeppelin-submarine-cpu:1.0.0, Comment: tensorflow_1.13.0-cpu & hadoop_3.1.2
zeppelin hadoop submarine gpu base image name: zeppelin-submarine-gpu:1.0.0, Comment: tensorflow_1.13.0-gpu & hadoop_3.1.2
```

#### Interpreter Docker Image:(Used to run the submarine interpreter in yarn)
```
zeppelin hadoop submarine Interpreter cpu image name: zeppelin-submarine-Interpreter-cpu:1.0.0, Comment: tensorflow_1.13.0-cpu & hadoop_3.1.2
zeppelin hadoop submarine Interpreter gpu image name: zeppelin-submarine-Interpreter-gpu:1.0.0, Comment: tensorflow_1.13.0-gpu & hadoop_3.1.2
```

#### Runtime Mount zeppelin lib
There is no need to pre-install zeppelin in docker image.
When zeppelin launches the submarine interpreter via `LaunchOnYarn`,
Will mount the zeppelin submarine lib jar into the docker container.
This very mode version of the update also reduces the size of the image

zjffdu commented on submarine/src/main/java/org/apache/zeppelin/submarine/IPySubmarineInterpreter.java:
```diff
@@ -0,0 +1,44 @@
+    PySubmarineInterpreter pySparkInterpreter =
+        getInterpreterInTheSameSessionByClassName(PySubmarineInterpreter.class, false);
+
+    pySubmarineInterpreter
```
`pySubmarineInterpreter` is never used ?

liuxunorg replied:
Currently we are using the `pySubmarineInterpreter`
Later, when we supported the `IpySubmarineInterpreter`, I will add it again.
I refer to the code of the `Zeppelin Python Interpreter`,
I am not familiar with `ipython` at present, the current code, It can be used normally,
Therefore, in order to stabilize the system, still retain this part of the code.

zjffdu commented on submarine/pom.xml:
```diff
@@ -0,0 +1,166 @@
+      </exclusions>
+    </dependency>
+    <dependency>
+      <groupId>com.hubspot.jinjava</groupId>
```
Can you check its licence ?

liuxunorg replied:
It's Apache License 2.0.
https://github.com/HubSpot/jinjava/blob/master/LICENSE

zjffdu commented:
Thanks @liuxunorg for this big contribution, I will merge it if no more comments. Since this PR is independent on other components, we can fix issues in other PRs later.

liuxunorg commented:
Thanks @zjffdu , @felixcheung  review code, **[Hadoop submarine project (https://hadoop.apache.org/submarine)](https://hadoop.apache.org/submarine)** is developing new features support `Pytroch`, `MXNet`, such as `model publish`, `model incremental update` and , etc. These features are ideal for integration in zeppelin.
I will continue to upgrade and update the hadoop submarine interpreter. :-)
